### ISSUE TITLE
ERROR: could not determine data type of parameter when projecting PostgreSQL UDT bind value

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#13174]
We have a few UDTs in our integration tests, which are used as "smoke tests". I.e. they combine various data types, and nested UDTs:

```sql
CREATE TYPE u_street_type AS (
  street VARCHAR(100),
  no VARCHAR(30),
  floors integer[],
  f_1323 bytea
);

CREATE TYPE u_address_type AS (
  street u_street_type,
  zip VARCHAR(50),
  city VARCHAR(50),
  country u_country,
  since DATE,
  code INTEGER,
  f_1323 bytea
);
```

So far, our integration tests mostly used binds of these types when inserting them into tables. But the following should work as well:

```java
UAddressTypeRecord c1 = new UAddressTypeRecord();
UAddressTypeRecord c2 = ctx.fetchValue(select(val(c1)));

assertEquals(c1, c2);
```

However, it doesn't:

```
org.jooq.exception.DataAccessException: SQL [select row(?, ?, ?, ?::"public"."u_country", cast(? as date), ?, ?) -- SQL rendered with a free trial version of jOOQ 3.17.0-SNAPSHOT]; ERROR: could not determine data type of parameter $1
	at org.jooq_3.17.0-SNAPSHOT.POSTGRES.debug(Unknown Source)
	at org.jooq.impl.Tools.translate(Tools.java:3106)
	at org.jooq.impl.DefaultExecuteContext.sqlException(DefaultExecuteContext.java:670)
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:355)
	at org.jooq.impl.AbstractResultQuery.fetchLazy(AbstractResultQuery.java:301)
	at org.jooq.impl.AbstractResultQuery.fetchLazyNonAutoClosing(AbstractResultQuery.java:322)
	at org.jooq.impl.SelectImpl.fetchLazyNonAutoClosing(SelectImpl.java:2845)
	at org.jooq.impl.ResultQueryTrait.fetchOne(ResultQueryTrait.java:509)
	at org.jooq.impl.Tools.attach(Tools.java:1338)
	at org.jooq.impl.DefaultDSLContext.fetchOne(DefaultDSLContext.java:4616)
	at org.jooq.impl.DefaultDSLContext.lambda$53(DefaultDSLContext.java:4636)
	at org.jooq.impl.Tools.attach(Tools.java:1338)
	at org.jooq.impl.DefaultDSLContext.fetchValue(DefaultDSLContext.java:4636)
	at org.jooq.test.PostgresTest.lambda$219(PostgresTest.java:4251)
	at org.jooq.test.jOOQAbstractTest.withDSLContext(jOOQAbstractTest.java:1962)
	at org.jooq.test.jOOQAbstractTest.bindAndInline(jOOQAbstractTest.java:1949)
	at org.jooq.test.PostgresTest.testPostgresUDTsWithNulls(PostgresTest.java:4249)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:93)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:529)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:756)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:452)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:210)
Caused by: org.postgresql.util.PSQLException: ERROR: could not determine data type of parameter $1
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2675)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2365)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:355)
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:490)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:408)
	at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:166)
	at org.postgresql.jdbc.PgPreparedStatement.execute(PgPreparedStatement.java:155)
	at org.jooq.tools.jdbc.DefaultPreparedStatement.execute(DefaultPreparedStatement.java:219)
	at org.jooq.tools.jdbc.DefaultPreparedStatement.execute(DefaultPreparedStatement.java:219)
	at org.jooq.impl.Tools.executeStatementAndGetFirstResultSet(Tools.java:4338)
	at org.jooq.impl.AbstractResultQuery.execute(AbstractResultQuery.java:236)
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:341)
	... 45 more
```

There are 2 problems here:

1. The top level `ROW` type does not have a type cast attached, it's an anonymous `ROW` type. That's wrong in case users want to pass that value to some function, or use it in unions, etc. The mapping to the UDT only happens in jOOQ in the client.
2. The `ROW` bind values aren't cast to the respective types, so PostgreSQL complains.

The cast might resolve both problems, to be investigated.

**Comment by lukaseder** [comment]
This issue was discovered while debugging another issue related to nested row types containing the `'null'` string value: https://github.com/jOOQ/jOOQ/issues/13169

**Comment by lukaseder** [comment]
Fixing this has shown yet another bug: https://github.com/jOOQ/jOOQ/issues/13175

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.17.0, 3.16.5 (https://github.com/jOOQ/jOOQ/issues/13179), and 3.15.9 (https://github.com/jOOQ/jOOQ/issues/13180)
